Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceDomGenerator.cpp

```diff
@@ -200,38 +202,57 @@ namespace AzToolsFramework
+            const Instance* topInstance = climbUpResult.m_reachedInstance;
+            for (auto instanceIter = climbedInstances.crbegin(); instanceIter != climbedInstances.crend(); ++instanceIter)
             {
+                // Create a path from the current top instance to the container entity of the given instance.
+                AZStd::string relativePathFromTopInstance =
+                    PrefabInstanceUtils::GetRelativePathBetweenInstances(*topInstance, instance);
+                const AZStd::string pathToContainerEntity =
+                    AZStd::string::format("%s/%s", relativePathFromTopInstance.c_str(), PrefabDomUtils::ContainerEntityName);
+                PrefabDomPath domPathToContainerEntity(pathToContainerEntity.c_str());
+                const PrefabDom& topInstanceTemplateDom =
+                    m_prefabSystemComponentInterface->FindTemplateDom((*topInstance).GetTemplateId());
                 // DOM reference can't be relied upon if the original DOM gets modified after reference creation.
                 // This scope is added to limit their usage and ensure DOM is not modified when it is being used.
-                const PrefabDom& rootTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom((*rootInstancePtr).GetTemplateId());
-                containerEntityDom.CopyFrom(*domPathToContainerEntity.Get(rootTemplateDom), instanceDom.GetAllocator());
+                {
+                    const auto* containerEntityDomPtr = domPathToContainerEntity.Get(topInstanceTemplateDom);
+                    if (containerEntityDomPtr)
+                    {
+                        containerEntityDom.CopyFrom(*containerEntityDomPtr, instanceDom.GetAllocator());
+                        break;
+                    }
+                }
+
+                topInstance = *instanceIter;
             }
 
-            // Sets the container entity DOM in the given instance DOM.
-            const AZStd::string containerEntityName = AZStd::string::format("/%s", PrefabDomUtils::ContainerEntityName);
-            PrefabDomPath containerEntityPath(containerEntityName.c_str());
-            containerEntityPath.Set(instanceDom, containerEntityDom, instanceDom.GetAllocator());
+            AZ_Assert(containerEntityDom.IsObject(), "Prefab - InstanceDomGenerator::UpdateContainerEntityInDomFromHighestAncestor - "
+                "Couldn't find container entity DOM in any ancestor.");
+            if (containerEntityDom.IsObject())
+            {
+                // Sets the container entity DOM in the given instance DOM.
+                const AZStd::string containerEntityName = AZStd::string::format("/%s", PrefabDomUtils::ContainerEntityName);
+                PrefabDomPath containerEntityPath(containerEntityName.c_str());
+                containerEntityPath.Set(instanceDom, containerEntityDom, instanceDom.GetAllocator());
+            }
```

Comment:
Looks like it is not optimized.

```cpp
PrefabDom containerEntityDom;   // has its own allocator
...
containerEntityDom.CopyFrom(*containerEntityDomPtr, instanceDom.GetAllocator()); // 1st copy
...
containerEntityPath.Set(instanceDom, containerEntityDom, instanceDom.GetAllocator()); // 2nd copy
```

You can probably do this:

```cpp
PrefabDom containerEntityDom(&(instanceDom.GetAllocator()));   // uses instanceDom's allocator
...
containerEntityDom.CopyFrom(*containerEntityDomPtr, containerEntityDom.GetAllocator()); // 1st copy
...
containerEntityPath.Set(instanceDom, containerEntityDom.Move()); // Move -> no copy
```

Alternatively, if the "Set" logic can be put right before `break`, we can do this:

```cpp
if (containerEntityDomPtr)
{
    PrefabDomValue containerEntityDom(*containerEntityDomPtr, 
        &(instanceDom.GetAllocator()));   // uses instanceDom's allocator -> 1st copy
    containerEntityPath.Set(instanceDom, containerEntityDom.Move()); // Move -> no copy

    // Or using one line
    containerEntityPath.Set(*containerEntityDomPtr, instanceDom.GetAllocator());  // 1st copy
}
```

Let me know if you have any question!

Reference: https://github.com/o3de/o3de/pull/13151#discussion_r1020537422

Reply:
@forkercat I updated this and rearranged so that code can be shared with GenerateContainerEntityDomFromRootInstance.